### PR TITLE
add new module: coin_market

### PR DESCRIPTION
**Pull request by lasers**
Hi, I converted @x86kernel's `coin_price` into `coin_market_cap` based on the site.

Later, I simplified the module name to `coin_market` instead. This get everything from data and turn them into placeholders... and also contain `limit` bit for the URL so it does not have to fetch everything (well over 700+ currencies) every 10 minutes.
* Example of using `limit`: https://api.coinmarketcap.com/v1/ticker/?limit=3
* Example of not using `limit`: https://api.coinmarketcap.com/v1/ticker/ 

Your `limit` number will be ~~high~~ low as possible to obtain all necessary `self.markets` (only once in `post_config_hook`) since there is no current way to obtain specified currency markets without making individual calls... which feel a bit like a mess way to do that. Site also leave this message. `Please limit requests to no more than 10 per minute.`

Also, the markets are ranked so if you're involved in 2 popular markets, your `limit` number might be `2`, `3`, `4`, `19`, or `444` (this one is def not popular).

The site offer prices in different currencies. Instead of using `convert` option, we can change currency in `format_coin` by using placeholders such as `{price_cad}`, `{price_gbp}`, etc.

~~I leave `color=degraded` in there to make it readable when using multiple markets. Many thanks.~~

**CoinMarketCap JSON API Documentation**: https://coinmarketcap.com/api/ 

lol
EDIT: Closes https://github.com/ultrabug/py3status/pull/938 😄

### CODE REVIEW

**Comment by ultrabug** [comment]
I'm all good here except that I get an error when running the module directly (self.first not defined)

Do you experience the same ?

File: py3status/modules/coin_market.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    {available_supply}    eg 16404825.0
+    {id}                  eg bitcoin
+    {last_updated}        eg 1498135152
+    {market_cap_usd}      eg 44119956596.0
+    {name}                eg Bitcoin
+    {percent_change_1h}   eg -0.17%
+    {percent_change_24h}  eg -1.93%
+    {percent_change_7d}   eg +14.73%
+    {price_btc}           eg 1.0
+    {price_usd}           eg 2689.45
```
Comment (tobes) [inline py3status/modules/coin_market.py]: Is there a reason for eg `{price_usd}` rather than just `{price}`, and how do I now set the currency?  In the original module this was possible

Comment (lasers) [inline py3status/modules/coin_market.py]: @ultrabug Got a log? I hadn't ran into this issue with Python `2`/`3`. 

@tobes By using `{price_gbp}` instead of `{price_usd}` -- Here's what `gbp` looks like https://api.coinmarketcap.com/v1/ticker/?convert=gbp&limit=5  -- You'll see that there are `price_gbp` along with `price_usd` too.

I didn't pick the `price_usd`. It came directly from the website. I just take the hand-off approach and pass everything (if possible) from raw data to py3status without touching.

The `format_coin` option should allow the users... 
1) To set their currency inside... instead of new currency option.
2) To set their currency symbol inside... instead of keeping the currency maps.
3) To set their currency formatting inside (ie if they wanted the symbol before or after the `{price_xxx}`... instead of ---- (We have nothing to deal with this).

Comment (tobes) [inline py3status/modules/coin_market.py]: >    To set their currency inside... instead of new currency option.
>    To set their currency symbol inside... instead of keeping the currency maps.

can the do that? what if I want EUR, GBP etc I don't think you can at the moment. If it got them all for me then that'd be nice but would be a big effort for not much gain?

It currently fails silently for `${price_usd} £{price_gbp}` but would work if they are reversed.

>    To set their currency formatting inside (ie if they wanted the symbol before or after the {price_xxx}... instead of ---- (We have nothing to deal with this).

yeah this is hard `€10` or `10€` can both be correct depending where you are.

Ok so this needs better error handling/documenting/example.

Comment (lasers) [inline py3status/modules/coin_market.py]: >can the do that? what if I want EUR, GBP etc I don't think you can at the moment. If it got them all for me then that'd be nice but would be a big effort for not much gain?

Can't. See the url above. It only can URL + `convert=%s`. One currency only unless we want to do multiple currencies + multiple calls. This currently works with `usd` + `any_other_currency`.

>It currently fails silently for ${price_usd} £{price_gbp} but would work if they are reversed.

The way it works right now, the `post_config_hook` part checks for the first 3 known possible `_currency` variables, then grab the first available one in the `format_coin`, break because of the answer above (We convert only 1).

>yeah this is hard €10 or 10€ can both be correct depending where you are.

I had currency map, then I thought, maybe it's better if we just leave them to the users instead... and to use mostly raw data (if possible). So bye, it goes.

Will add example of using gbp.

Comment (lasers) [inline py3status/modules/coin_market.py]: Really, what happened... If you compared https://api.coinmarketcap.com/v1/ticker/?convert=gbp&limit=1 to the placeholder list ( https://api.coinmarketcap.com/v1/ticker/?limit=1 )... There are extra `_usd` placeholders to all converted currencies... `{price_usd}`, `{24h_volume_usd}` and `{market_cap_usd}`. I didn't notice that before.

File: py3status/modules/coin_market.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    To view placeholders in different currency, replace `_usd` with a valid option
+    (eg '{price_usd}' to '{price_xxx}'). Case insensitive. Valid options are: AUD,
+    BRL, CAD, CHF, CNY, EUR, GBP, HKD, IDR, INR, JPY, KRW, MXN, RUB, otherwise USD.
+
+Examples:
+```
+# view coins in GBP
+coin_market {
+    format_coin = '[\?color=degraded {name}] {price_gbp:.2f}£ {percent_change_24h}'
```
Comment (tobes) [inline py3status/modules/coin_market.py]: `£{price_gbp:.2f}` please

File: py3status/modules/coin_market.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    {rank}                eg 1
+    {symbol}              eg BTC
+    {total_supply}        eg 16404825.0
+
+    Placeholders are retrieved directly from the URL.
+    The list was harvested only once and should not represent a full list.
+
+    To view placeholders in different currency, replace `_usd` with a valid option
+    (eg '{price_usd}' to '{price_xxx}'). Case insensitive. Valid options are: AUD,
+    BRL, CAD, CHF, CNY, EUR, GBP, HKD, IDR, INR, JPY, KRW, MXN, RUB, otherwise USD.
```
Comment (tobes) [inline py3status/modules/coin_market.py]: > Case insensitive

Do these not need to be lower case?  I think they do for the placeholders to work

File: py3status/modules/coin_market.py (hunk 1 above)
Comment (tobes) [inline py3status/modules/coin_market.py]: yeah it always gives usd.  So we can skip that.  Anyhow being clear to the users is important.

One observation is that doing the work in `post_config_hook()` will delay all modules from starting currently, because they all get initialized before any are started.  That can maybe be improved but it is a simple way of ensuring all modules don't have dependency issues.  But if it's not affecting you too much no problem

Comment (lasers) [inline py3status/modules/coin_market.py]: >It currently fails silently for ${price_usd} £{price_gbp} but would work if they are reversed.

Can you try again now w/ `new state (2)` ?

File: py3status/modules/coin_market.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                break
+
+        self.coin_data = self._get_coin_data()
+        limit = 0
+        for market_id in self.markets:
+            index = next((i for (
+                i, d) in enumerate(self.coin_data) if d['symbol'] == market_id), -1)
+            if index > limit:
+                limit = index
+        self.url = self.url + '&limit=%s' % (limit + 1)
```
Comment (lasers) [inline py3status/modules/coin_market.py]: Should I up this limit? Right now, it's perfect... Get everything we need, but because the markets are ranked, the limit number could change.

We follow 11th and 16th. 11th becomes 9th and/or 16th becomes 17th... then in the loop, the last market (from `self.markets`) could be same as the last 2nd market (from `self.markets`).

Maybe I should just wait until somebody report this first. Solution could be adding `try/except (new call)` automatically in the loop... something similar to #924.

File: py3status/modules/coin_market.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    format = '{format_coin}'
+    format_coin = '[\?color=degraded {name}] ${price_usd:.2f} {percent_change_24h}'
+    format_separator = ' '
+    markets = ['btc']
+    request_timeout = 5
+
+    def post_config_hook(self):
+        self.first = True
+        if not isinstance(self.markets, list):
+            self.markets = self.markets.split(',')
```
Comment (lasers) [inline py3status/modules/coin_market.py]: Easier to type `markets = 'btc, xre'` over `markets = ['btc','xre']`.

**Comment by ultrabug** [comment]
Error I get

```
Traceback (most recent call last):
  File "coin_market.py", line 165, in <module>
    module_test(Py3status)
  File "/home/ultrabug/venv/py3status/lib/python2.7/site-packages/py3status/module_test.py", line 35, in module_test
    print(getattr(module, method)(*method_args))
  File "coin_market.py", line 125, in coin_market
    if not self.first:
AttributeError: Py3status instance has no attribute 'first'
```

File: py3status/modules/coin_market.py (hunk 4 above)
Comment (tobes) [inline py3status/modules/coin_market.py]: Maybe just do a fresh call of them all if you can't find all the required currencies (I'm going to assume that the order changes rarely.  So move that code out of post config into it's own function etc

**Comment by lasers** [comment]
@ultrabug I think you're skipping `post_config_hook` entirely and I recall @tobes mentioned that it can happen in the terminal. Not sure what the issue exactly was.

**Comment by tobes** [comment]
@ultrabug 

> Error I get

I can't replicate this error either by running in test mode or as an active module

@lasers 

I think it would be good to add some comments to this code.  eg what is special about the first run.  I shouldn't need to have to read all the code to know what is happening or why.  Once when I was young I thought that comments were a waste of time but as time has gone on I realize why they are great (when done well) commenting is a useful art.

**Comment by ultrabug** [comment]
It's ok for me, tell me when we're good here, so much for inception conversations

File: py3status/modules/coin_market.py (hunk 5 above)
Comment (maximbaz) [inline py3status/modules/coin_market.py]: Honestly I'd recommend staying away from creating multiple ways to define the same thing, it's not that more difficult to type a list, and you have to do it quite rarely 🙂 Your docstring is confusing (what I thought when I read "comma-separated list": mmm... as in comma-separated string? or a real python list? and then I found this line). Also if ever this has to be deprecated, you'll have to deprecate two ways of defining the same thing 🙂

**Comment by lasers** [comment]
I'd like to fix this first so it's done in the loop instead of `post_config_hook` then all modules won't have to wait on this one... since I recognize the delayed startup. I'm in this to optimize things a bit more.

Few days ago, I sent a Google Form request for something like `?coin=btc&ltc&xre` to give us data for 3 coins... instead of using `?limit=32` to get 32 coins and output 3 coins... and `?convert=gbp&cad` to give us coins in both data. No idea if he would do it. :-)

File: py3status/modules/coin_market.py (hunk 5 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Fixed. 👍

File: py3status/modules/coin_market.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+                temporary = {}
+                for k, v in market.items():
+                    if 'percent_change' in k and v:
+                        if float(v) >= 0:
+                            _fmt = '\?color=good %s' % v
+                        else:
+                            _fmt = '\?color=bad %s' % v
+                        _fmt = '+' + _fmt if float(v) > 0 else _fmt
+                        temporary[k] = self.py3.safe_format(_fmt + '%')
```
Comment (tobes) [inline py3status/modules/coin_market.py]: Using a threshold might work better.  It would allow the user to have greater control and may well simplify things.

You really should add some comments.

File: py3status/modules/coin_market.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                    else:
+                        temporary[k] = v
+
+                data.append(self.py3.safe_format(self.format_coin, temporary))
+
+        format_separator = self.py3.safe_format(self.format_separator)
+        format_coin = self.py3.composite_join(format_separator, data)
+
+        if self.first:
```
Comment (tobes) [inline py3status/modules/coin_market.py]: can this ever be true?

File: py3status/modules/coin_market.py (hunk 6 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Do threshold work with negative numbers? I can't be sure that it does... I think `weather_owm` could have this same issue too now I think about it.

File: py3status/modules/coin_market.py (hunk 7 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: I didn't realize I pushed `coin_market: update limit automatically` until this review. Sorry. It's still `wip`. Gotta sleep on it first to see how we can make everything works together well in this module.

File: py3status/modules/coin_market.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    {'full_text': u' $2701.70 '},
+    {'color': '#FF0000', 'full_text': u'-0.42%'},
+]
+"""
+
+
+class Py3status:
+    cache_timeout = 600
+    format = '{format_coin}'
+    format_coin = '[\?color=degraded {name}] ${price_usd:.2f} {percent_change_24h}'
```
Comment (lasers) [inline py3status/modules/coin_market.py]: Heh. No abuse there. The idea was that if we have something like this in... then the users could just change via `color_xxxx` without touching the format. 

Another idea is to get rid of `color_xxxx` and have users do it via `format` -- Not friendly. Sometimes users just want to change one thing without hoarding everything. Idk.

Comment (lasers) [inline py3status/modules/coin_market.py]: Done. `s/degraded/#ff0`

File: py3status/modules/coin_market.py (hunk 6 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Done. `s/color=xxxxx/thresholds` 
Done. Some comments.

File: py3status/modules/coin_market.py (hunk 7 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Done. `coin_market: wtf state`.... with `self.first` for faster `py3status` startup and `self.first_check` for smart `self.markets` error checking.

File: py3status/modules/coin_market.py (hunk 8 above)
Comment (maximbaz) [inline py3status/modules/coin_market.py]: I'd argue that this is too colorful for a default, maybe make something less sophisticated and let users play with colors if they want to? 

I set this for myself for example: `format_coin = '{symbol} ${price_usd:.2f}'`

**Comment by maximbaz** [comment]
Thank you for fixing the startup speed 🙂 I'm using this module for a few days now, works nicely!

Can't wait for a new release with this module included 😉

File: py3status/modules/coin_market.py (hunk 8 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: You can try `limit = 10` for a while... to understand why I wanted to add colors. It can be a tad hard to parse things without them. This is to make multiple markets readability on the off chance somebody could be using them and that they don't have to configure `format` out of the box... assuming default `format` is not awful.

```
# color just works without adding anything
coin_markets {
    limit = 10
}
```
```
# color just works without adding anything
coin_markets {
    markets = ['btc', 'dash', 'eth', 'sky'] 
}
```

Comment (maximbaz) [inline py3status/modules/coin_market.py]: Well there certainly are ways to make it distinguishable without introducing many colors by default 🙂 I'm fine with keeping it as it is, just saying, when I installed this module my first impression was "wow, what a rainbow 🌈" 😄 

```
coin_markets {
    markets = ['btc', 'dash', 'eth', 'sky'] 
}
```

image


```
coin_market {
  markets = ['btc', 'dash', 'eth', 'sky']
  format_coin = '{symbol} ${price_usd:.2f}'
  format_separator = '   '
}
```

image

File: py3status/modules/coin_market.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+"""
+Display cryptocurrency data from https://coinmarketcap.com
+
+Configuration parameters:
+    cache_timeout: refresh interval for this module. A message from the site:
+        Please limit requests to no more than 10 per minute.
+        (default 600)
+    format: display format for this module (default '{format_coin}')
+    format_coin: display format for coins
+        (default '[\?color=#ff0 {name}] ${price_usd:.2f} {percent_change_24h}')
```
Comment (tobes) [inline py3status/modules/coin_market.py]: please kill this color stuff it makes me want to cry.

File: py3status/modules/coin_market.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+
+    Placeholders are retrieved directly from the URL.
+    The list was harvested only once and should not represent a full list.
+
+    To print coins in different currency, replace or replicate the placeholders
+    below with a valid option (eg '{price_gbp}') to create additional placeholders:
+
+    {price_xxx}           eg 1819.67786724  # GBP
+    {24h_volume_xxx}      eg 668740360.049  # GBP
+    {market_cap_xxx}      eg 29926922616.0  # GBP
```
Comment (tobes) [inline py3status/modules/coin_market.py]: why use `xxx` but say output is in `GBP`? a real example is more helpful that one that would actually give bad output

File: py3status/modules/coin_market.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    thresholds = [(-100, 'bad'), (0, 'good')]
+
+    def post_config_hook(self):
+        self.first_run = True
+        self.first_use = True
+        self.convert = None
+        self.url = self.reset_url = 'https://api.coinmarketcap.com/v1/ticker/'
+        self.limit = None
+
+        # ranked markets or cherrypicked markets?
```
Comment (tobes) [inline py3status/modules/coin_market.py]: this comment asks a question rather than clarifying things.  Maybe I'm just confused here.

To be honest I don't like this overloading of parameters you are doing.

File: py3status/modules/coin_market.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+                self.url = self.reset_url = self.reset_url + self.convert
+                break
+
+        # create '(?|&)limit'
+        if self.limit:
+            self._update_limit(None)
+
+    def _get_coin_data(self, reset_url=False):
+        if reset_url:
+            self.url = self.reset_url
```
Comment (tobes) [inline py3status/modules/coin_market.py]: this logic feels nasty.  sometimes reset_url is url sometimes boolean why?

File: py3status/modules/coin_market.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+                    data) if d['symbol'] == market_id), -1)
+                if index >= limit:
+                    limit = index
+                    limit += 5  # padding
+
+        self.url += 'limit=%s' % limit
+
+    def _strip_data(self, data):
+        # we don't strip if self.limit. otherwise, we reduce 997 coins
+        # to 5 coins by stripping everything not in the markets.
```
Comment (tobes) [inline py3status/modules/coin_market.py]: this comment is very specific about numbers but does not seem to limit to 5 coins

File: py3status/modules/coin_market.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+        return new_data
+
+    def _organize_data(self, data):
+        # compare len(markets) w/ len(stripped(997 coins)).
+        new_data = self._strip_data(data)
+        is_equal = len(new_data) == len(self.markets)
+
+        # first time bad? the user entered bad markets. stop here.
+        # └── else, we make a limit based on 997 coins.
```
Comment (tobes) [inline py3status/modules/coin_market.py]: `└──` why?

comments are to help us not add confusion, why this 997 coins am I missing something?

File: py3status/modules/coin_market.py (hunk 9 above)
Comment (tobes) [inline py3status/modules/coin_market.py]: if it was at least a named color then the user could change it easily but I think it would still be hard to implement cleanly

File: py3status/modules/coin_market.py (hunk 14 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: >this comment is very specific about numbers but does not seem to limit to 5 coins

>comments are to help us not add confusion, why this 997 coins am I missing something?

An example. We really do parse `997 coins`... then work with `%s coins`.

File: py3status/modules/coin_market.py (hunk 12 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Will change one.

File: py3status/modules/coin_market.py (hunk 11 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Do the user want `ranked markets` or `cherrypicked markets`? We find out soon.
It's either `markets = 10` or `markets = ['btc', 'ltc', 'doge']` in the config.

>To be honest I don't like this overloading of parameters you are doing.
I create many as I need to work with startup optimization, first time checking (to diff between bad user inputted markets and markets out of the limit) along with diff urls, etc.

File: py3status/modules/coin_market.py (hunk 10 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: We only can change those three placeholders. I will change `eg` strings.

File: py3status/modules/coin_market.py (hunk 9 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: I think it'd be nice to import 16 colors from `~/.Xresources`. It'd be `color=3` for yellow, then override colors from `color_*` inside the modules by just `color_3 = #75507b`. No giving names like `degraded` because it's not degraded here... and `color_yellow = #75507b` wouldn't make sense either (it's purple hex color).
```
format_coin = [\?color=3 {name}] ${price_usd:.2f} {percent_change_24h}
color3 = '#75507b'
```

`rofi` do something like this too in `~/.Xresources`.
```
! breeze-dark
rofi.color-enabled: true
rofi.color-window: #31363B, #002b36, #31363B
rofi.color-normal: #31363B, #ffffff, #31363B, #ffffff, #31363B
rofi.color-active: #31363B, #268bd2, #31363B, #268bd2, #ffffff
rofi.color-urgent: #31363B
```
If no color in `~/.Xresources`, we use default `linux/tango` color theme.

File: py3status/modules/coin_market.py (hunk 14 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: https://api.coinmarketcap.com/v1/global/

If I got this right, `828` active currencies + `169` dead/stalled currencies = `997` currencies on every update... Crazy. It probably only go up from here. Good thing there is a `limit` to reduce data overhead... but if `coinmarketapi.com` website updates with new way to obtain coins directly, then we can reduce lot of parameters.

**Comment by lasers** [comment]
Done. I updated things to meet your criteria.

File: py3status/modules/coin_market.py (hunk 11 above)
Comment (maximbaz) [inline py3status/modules/coin_market.py]: I would be very interested to hear who is using `markets = 10` and why, I cannot imagine a scenario why would someone monitor prices of who-knows-which coins - but I might be wrong :)

**Comment by maximbaz** [comment]
I noticed that the sorting of `markets` parameter is ignored, i.e. I set `markets = ['ltc', 'btc']` but BTC is still the first one being outputted. @lasers is it possible to preserve the order?

**Comment by lasers** [comment]
@maximbaz Fixed. The markets are ranked. I figure it'd be nice to toss in an unordered list of markets and have them all lined up accordingly. Bitcoin would be `No. 1` for long long long time. Majority of the users probably won't dabble in a large list of markets so we just follow what the users put in. 

For ordered markets, use `limit = 5` (ranked) as it should be.

@tobes @ultrabug Let me know if you want me to squash the commits and/or give me some criticism. @maximbaz have been my generous guinea pig for weeks and I take it he don't find any problems with this module.

**Comment by maximbaz** [comment]
Everything works great indeed 😉 Thanks for the fix 👍

**Comment by tobes** [comment]
I think we can do better with the thresholds.

the following diff allows us to use the thresholds more flexibly so we can have formats like

`\?color=1h {name} {price_usd}` 

the documentation would also need updating a little to reflect this.

```diff
diff --git a/py3status/modules/coin_market.py b/py3status/modules/coin_market.py
index 736599c..c7ffd12 100644
--- a/py3status/modules/coin_market.py
+++ b/py3status/modules/coin_market.py
@@ -7,7 +7,7 @@ Configuration parameters:
         Please limit requests to no more than 10 per minute. (default 600)
     format: display format for this module (default '{format_coin}')
     format_coin: display format for coins
-        (default '{name} ${price_usd:.2f} {percent_change_24h}')
+        (default '{name} ${price_usd:.2f} [\?color=24h {percent_change_24h}]')
     format_separator: show separator only if more than one (default ' ')
     markets: number of top ranked markets or list of user-inputted markets
         (default ['btc'])
@@ -74,7 +74,7 @@ losers
 class Py3status:
     cache_timeout = 600
     format = '{format_coin}'
-    format_coin = '{name} ${price_usd:.2f} {percent_change_24h}'
+    format_coin = '{name} ${price_usd:.2f} [\?color=24h {percent_change_24h}]'
     format_separator = ' '
     markets = ['btc']
     request_timeout = 5
@@ -177,11 +177,11 @@ class Py3status:
             temporary = {}
             for k, v in market.items():
                 if 'percent_change' in k and v:
-                    _fmt = '+%s%%' % v if float(v) > 0 else '%s%%' % v
-                    temporary[k] = self.py3.safe_format('\?color=%s %s' % (
-                        self.py3.threshold_get_color(v, k), _fmt))
-                else:
-                    temporary[k] = v
+                    # set the thresholds for this market
+                    # these will be colors 1h 24h and 7d
+                    threshold_name = k[15:]
+                    self.py3.threshold_get_color(v, threshold_name)
+                temporary[k] = v
 
             new_data.append(self.py3.safe_format(self.format_coin, temporary))
```

File: py3status/modules/coin_market.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        return new_data
+
+    def _manipulate_data(self, data):
+        # we mess with raw data to get the results we want. we're only
+        # adding colors + symbols to all percent placeholders.
+        new_data = []
+        for market in data:
+            temporary = {}
+            for k, v in market.items():
+                if 'percent_change' in k and v:
```
Comment (tobes) [inline py3status/modules/coin_market.py]: being paranoid maybe update this to `if k.startswith('percent_change_') and v:`

Comment (lasers) [inline py3status/modules/coin_market.py]: I think paranoid. I want to leave it as-is and wait until the shit hits the fan first.

Comment (tobes) [inline py3status/modules/coin_market.py]: ok how about `if 'percent_change_' in k and v:` that way `k[15:]` will never fail

File: py3status/modules/coin_market.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
+"""
+Display cryptocurrency data from https://coinmarketcap.com
+
```
Comment (tobes) [inline py3status/modules/coin_market.py]: a bit more explanation would be good here.

> Data such as price, percentage change, and total supply for a wide range of crypto currencies.  Prices can be shown in traditional currencies eg YEN, EUR.

Comment (lasers) [inline py3status/modules/coin_market.py]: I explained. 👍

File: py3status/modules/coin_market.py (hunk 15 above)
Comment (lasers) [inline py3status/modules/coin_market.py]: Okay. :-)

**Comment by ultrabug** [comment]
Thanks a lot @lasers !
